core/src/main/java/org/jdbi/v3/core/mapper/ColumnMappers.java

```diff
@@ -213,15 +212,18 @@ public <T> Optional<ColumnMapper<T>> findFor(QualifiedType<T> type) {
             return cached;
         }
 
-        Optional<ColumnMapper<T>> mapper = (Optional) factories.stream()
-                .flatMap(factory -> JdbiOptionals.stream(factory.build(type, registry)))
-                .findFirst();
-
-        mapper.ifPresent(m -> m.init(registry));
-
-        cache.put(type, mapper);
+        for (QualifiedColumnMapperFactory factory : factories) {
```

Comment:
this seems to be an excellent opportunity to fix some technical debt from the past (this will go into main branch):

```java
    @SuppressWarnings({ "unchecked" })
     public <T> Optional<ColumnMapper<T>> findFor(QualifiedType<T> type) {
        return Optional.ofNullable((ColumnMapper<T>) cache.computeIfAbsent(type, t -> {
            for (QualifiedColumnMapperFactory factory : factories) {
                ColumnMapper<?> mapper = factory.build(t, registry).orElse(null);
                if (mapper != null) {
                    mapper.init(registry);
                    return mapper;
                }
            }
            return null;
        }));
```

and change the cache to no longer hold the optionals but only the actual mappers (which makes the cache smaller because it no longer needs to cache negative hits).

Reply:
ok, after some testing, removing the negative cache is a bad idea. One ends up with an infinite loop when resolving e.g. a map type. So keep the optional. :-) It still makes the code nicer.

Reply:
I think your infinite looping might be re-discovering this `computeIfAbsent` behavior, rather than anything to do with the negative cache. It might still be possible to remove the negative caching, although I am not sure it hurts or helps much, since missing mappers should be rare (usually throws an exception)